XMonad/Prelude.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE BangPatterns #-}
+--------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Prelude
+-- Copyright   :  slotThe <soliditsallgood@mailbox.org>
+-- License     :  BSD3-style (see LICENSE)
+--
+-- Maintainer  :  slotThe <soliditsallgood@mailbox.org>
+--
+-- Utility functions and re-exports for a more ergonomic developing
+-- experience.  Users themselves will not find much use here.
+--
+--------------------------------------------------------------------
+module XMonad.Prelude (
+    module Exports,
+    fi,
+    chunksOf,
+    (.:),
+    (!?),
+) where
+
+import Control.Applicative as Exports
+import Control.Monad       as Exports
+import Data.Bool           as Exports
+import Data.Char           as Exports
+import Data.Foldable       as Exports
```

Comment:
This is going to collide with `Data.List`, and it's not clear to me which one "wins" (there "should" be a multiple definition error in the export line, but the way GHC checks things, there won't be). This shouldn't affect behavior but may affect performance in some cases, since Foldable is more general and therefore less optimized… although they're lists, so they're not going to perform especially well anyway.

Come to think of it, if `Data.List` wins, that could cause some breakage for things expecting `Data.Foldable` to work on e.g. `Set`. But I assume CI would catch that.

Reply:
Come to think of it, `Data.List` these days is mostly reexports from `Data.Foldable` so the above collision can't really occur (or can, but harmlessly). I still wonder if having both is safe in the long term.

Reply:
Yeah most of the things that are applicable to both lists and foldables is just re-exported from `Data.Foldable` now.  In general I think most of the `base` libraries work with each other more often than not.  There are certainly notable exceptions (like `Control.Category` having an instance of `(.)` that's incompatible with `base`s) but since a good number of modules simply import `XMonad.Prelude` unqualified and `xmonad-contrib` still compiles, everything seems fine here.

There's also the fact that `base` only changes with new versions of GHC, so if something were to break then I suspect we'd be the first to know and could fix things before users upgrade.

Reply:
Found one more: `X.P.Shell.split` is something like `Data.List.Split.splitOn . (:[])` with a special case for `[]`, returning `[]` instead of `[""]`.